src/EditorFeatures/Core/Tagging/AsynchronousTaggerProvider.Tagger.cs

```diff
@@ -0,0 +1,257 @@
+            }
+
+            private void OnResumed(object sender, EventArgs e)
+            {
+                _batchChangeNotifier.Resume();
+            }
+
+            private void OnTagsChangedForBuffer(ICollection<KeyValuePair<ITextBuffer, NormalizedSnapshotSpanCollection>> changes)
+            {
+                // Note: This operation is uncancellable. Once we've been notified here, our cached tags
+                // in the tag source are new. If we don't update the UI of the editor then we will end
+                // up in an inconsistent state between us and the editor where we have new tags but the
+                // editor will never know.
+
+                _tagSource.RegisterNotification(() =>
+                {
+                    foreach (var change in changes)
+                    {
+                        if (change.Key != _subjectBuffer)
+                        {
+                            continue;
+                        }
+
+                        // Now report them back to the UI on the main thread.
+                        _batchChangeNotifier.EnqueueChanges(change.Value);
+                    }
+
+                }, TaggerDelay.NearImmediate.ComputeTimeDelayMS(), CancellationToken.None);
+            }
+
+            public IEnumerable<ITagSpan<TTag>> GetTags(NormalizedSnapshotSpanCollection requestedSpans)
+            {
+                return GetTagsWorker(requestedSpans, accurate: false, cancellationToken: CancellationToken.None);
+            }
+
+            public IEnumerable<ITagSpan<TTag>> GetAllTags(NormalizedSnapshotSpanCollection requestedSpans, CancellationToken cancellationToken)
+            {
+                return GetTagsWorker(requestedSpans, accurate: true, cancellationToken: cancellationToken);
+            }
+
+            private IEnumerable<ITagSpan<TTag>> GetTagsWorker(
+                NormalizedSnapshotSpanCollection requestedSpans,
+                bool accurate,
+                CancellationToken cancellationToken)
+            {
+                if (requestedSpans.Count == 0)
+                {
+                    return SpecializedCollections.EmptyEnumerable<ITagSpan<TTag>>();
+                }
+
+                var buffer = requestedSpans.First().Snapshot.TextBuffer;
+                var tags = accurate
+                    ? _tagSource.GetAccurateTagIntervalTreeForBuffer(buffer, cancellationToken)
+                    : _tagSource.GetTagIntervalTreeForBuffer(buffer);
+
+                if (tags == null)
+                {
+                    return SpecializedCollections.EmptyEnumerable<ITagSpan<TTag>>();
+                }
+
+                var result = GetTags(requestedSpans, tags);
+
+                DebugVerifyTags(requestedSpans, result);
+
+                return result;
+            }
+
+            private static IEnumerable<ITagSpan<TTag>> GetTags(NormalizedSnapshotSpanCollection requestedSpans, TagSpanIntervalTree<TTag> tags)
```

Comment:
I would appreciate a reduction of methods called "GetTags" in this class -- maybe call this FilterTagsToRequestedSpans or something?